packages/contracts-bedrock/src/L2/L2ToL2CrossDomainMessenger.sol

```diff
@@ -104,16 +117,61 @@ contract L2ToL2CrossDomainMessenger is IL2ToL2CrossDomainMessenger, ISemver, Tra
         if (_target == Predeploys.CROSS_L2_INBOX) revert MessageTargetCrossL2Inbox();
         if (_target == Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert MessageTargetL2ToL2CrossDomainMessenger();
 
+        /*
         bytes memory data = abi.encodeCall(
-            L2ToL2CrossDomainMessenger.relayMessage,
+            L2ToL2CrossDomainMessenger.execute,
             (_destination, block.chainid, messageNonce(), msg.sender, _target, _message)
         );
         assembly {
             log0(add(data, 0x20), mload(data))
         }
+         */
+
+        emit SentMessage(_destination, _target, messageNonce(), msg.sender, _message);
+
         msgNonce++;
     }
 
+    function relayMessage2(ICrossL2Inbox.Identifier calldata _id, bytes calldata _sentMessage) external payable nonReentrant {
+        // Ensure the log came from the messenger and validate with the cross l2 inbox
+        if (_id.origin != Predeploys.L2_TO_L2_CROSS_DOMAIN_MESSENGER) revert("not the l2 cdm");
+        CrossL2Inbox(Predeploys.CROSS_L2_INBOX).validateMessage(_id, keccak256(_sentMessage));
+
+        // Open Question:
+        //   - should we assert length on `_sentMessage` prior to decoding? It's structure should implicitly be known and valid
```

Comment:
I think that its fine to not assert on the length, we aren't accepting inputs from an untrusted contract, we know the only way that the event can be emitted